include/nbl/builtin/hlsl/tgmath/impl.hlsl

```diff
@@ -203,6 +208,42 @@ struct erf_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScala
 	}
 };
 
+// logn-gamma function
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T>)
+struct lgamma_helper<T NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+{
+	// implementation from Numerical Recipes in C, 2nd ed.
+    static T __call(T val)
+    {
+        T x, y;
```

Comment:
at the very start of the function you need to do the following
```cpp
const T thresholds[3] = {float16 threshold, float32 threshold, float64 threshold};
if (thresholds[findLSB(sizeof(T))]<val)
   return +INF;
```

there should also be a check for value being negative and reaching particular values, but every odd value (and fractions below it) would produce negative gamma and a NaN though log anyway.